### PR TITLE
Add Unit Tests for IPortfolioOptimizer Classes

### LINKED ISSUE #4494: Add Unit Tests for IPortfolioOptimizer Classes
#### Expected Behavior
Unit tests cover `IPortfolioOptimizer` implementations.

#### Actual Behavior
No unit test for `IPortfolioOptimizer` implementations.

#### Potential Solution
Add unit tests.

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)
- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by wtindall1**
#### Description

Added unit tests for IPortfolioOptimizer implementations:
- MaximumSharpeRatioPortfolioOptimizer
- MinimumVariancePortfolioOptimizer
- UnconstrainedMeanVariancePortfolioOptimizer
- PortfolioOptimizerPythonWrapper

#### Related Issue

 Closes #4494 

#### Motivation and Context

The implementations were not covered by unit tests.

#### Requires Documentation Change

No

#### How Has This Been Tested?

Ran the new and existing tests.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Tests/Algorithm/Framework/Portfolio/MaximumSharpeRatioPortfolioOptimizerTests.cs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    {
+        private Dictionary<int, double[,]> _historicalReturns = new();
+        private Dictionary<int, double[]> _expectedReturns = new();
+        private Dictionary<int, double[,]> _covariances = new();
+        private Dictionary<int, double[]> _expectedResults = new();
+
+        [OneTimeSetUp]
+        public void Setup()
+        {
+            double[,] historicalReturns1 = new double[,] { { 0.02, -0.02, 0.28 }, { -0.50, -0.29, -0.13 }, { 0.81, 0.29, 0.31 }, { -0.03, -0.00, 0.01 } };
```
Comment (jhonabreul) [inline Tests/Algorithm/Framework/Portfolio/MaximumSharpeRatioPortfolioOptimizerTests.cs:35]: In Lean it is a common practice to use prioritize the use of `var` vs. explicitly typing the variable

File: Tests/Algorithm/Framework/Portfolio/MaximumSharpeRatioPortfolioOptimizerTests.cs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+namespace QuantConnect.Tests.Algorithm.Framework.Portfolio
+{
+    [TestFixture]
+    public class MaximumSharpeRatioPortfolioOptimizerTests
+    {
+        private Dictionary<int, double[,]> _historicalReturns = new();
+        private Dictionary<int, double[]> _expectedReturns = new();
+        private Dictionary<int, double[,]> _covariances = new();
+        private Dictionary<int, double[]> _expectedResults = new();
```
Comment (jhonabreul) [inline Tests/Algorithm/Framework/Portfolio/MaximumSharpeRatioPortfolioOptimizerTests.cs:30]: The keys are just indices, perhaps a list or array would be enough?

**Review by jhonabreul — CHANGES_REQUESTED** [review]
Thanks for your contribution 🚀 Just leaving a couple minor comments

**Review by jhonabreul — APPROVED** [review]
Thank you!

File: Tests/Algorithm/Framework/Portfolio/MinimumVariancePortfolioOptimizerTests.cs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+            Assert.AreEqual(_expectedResults[testCaseNumber], result.Select(x => Math.Round(x, 6)));
+            Assert.AreEqual(1d, result.Select(x => Math.Round(Math.Abs(x), 6)).Sum());
+        }
+
+        [TestCase(4)]
+        [TestCase(5)]
+        [TestCase(6)]
+        [TestCase(7)]
+        public void TestOptimizeWeightingsSpecifyingTargetReturns(int testCaseNumber)
```
Comment (Martin-Molinero) [inline Tests/Algorithm/Framework/Portfolio/MinimumVariancePortfolioOptimizerTests.cs:130]: Let's remove `Test` from the test names, same comment for the other tests & files 👍

File: Tests/Algorithm/Framework/Portfolio/MinimumVariancePortfolioOptimizerTests.cs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        {
+            var testOptimizer = new MinimumVariancePortfolioOptimizer();
+            var historicalReturns = new double[,] { { } };
+
+            var result = testOptimizer.Optimize(historicalReturns);
+
+            Assert.AreEqual(Array.Empty<double>(), result);
+        }
+
+        public void SingleSecurityPortfolioReturnsOne()
```
Comment (Martin-Molinero) [inline Tests/Algorithm/Framework/Portfolio/MinimumVariancePortfolioOptimizerTests.cs:187]: These are missing the `[Test]` attribute, same for `UnconstrainedMeanVariancePortfolioOptimizer`

**Review by Martin-Molinero — COMMENTED** [review]
Thank you @wtindall1!
Leaving a couple of comments. It seems there's room to create a base `PortfolioOptimizerTest` class which defines shared cases to be tested and each override can just define the result it produces.
If your interested think we should add some tests for the `RiskParityPortfolioOptimizer` & the `PortfolioOptimizerPythonWrapper`

**Comment by wtindall1** [comment]
Hi @Martin-Molinero - thanks for the review, I will address your comments. For the `RiskParityPortfolioOptimizer` , is this not covered by the `RiskParityPortfolioOptimizerTests` class, or is there additional coverage you were hoping for?

**Comment by Martin-Molinero** [comment]
> For the RiskParityPortfolioOptimizer , is this not covered by the RiskParityPortfolioOptimizerTests class

Oh nice, sorry didn't realize 👍 existing tests look good. For `PortfolioOptimizerPythonWrapper` can take a look for inspiration at `BrokerageModelTests` basically testing the wrapping functionality of a python wrapper

**Review by Martin-Molinero — APPROVED** [review]
Nice work @wtindall1! Thank you 🚀

**Pull request merged**